Web Inspector: Regression(260175@main) Grid overlay for grids using repeat syntax triggers assert performing downcast

patrickangle opened the pull request:
#### a465a8a38ab52cfd57d2c2e710d72ea6f53d9c81
<pre>
Web Inspector: Regression(260175@main) Grid overlay for grids using repeat syntax triggers assert performing downcast
<a href="https://bugs.webkit.org/show_bug.cgi?id=255311">https://bugs.webkit.org/show_bug.cgi?id=255311</a>
rdar://107764654

Reviewed by Devin Rousso.

After 260175@main, CSSGridIntegerRepeatValue and CSSGridAutoRepeatValue no longer inherit from CSSValueList, and instead
inherit from CSSValueContainingVector. The inspector overlay code was written in such a way that we were checking if the
class was one of the repeat types, but then downcasting to CSSValueList making that relation less apparent. Converting
the downcasts to dynamicDowncasts where we previously had the `is` checks fixes the issue and guards against this
mistake in the future.

* LayoutTests/inspector/dom/showGridOverlay.html:
* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::authoredGridTrackSizes):

Canonical link: <a href="https://commits.webkit.org/262869@main">https://commits.webkit.org/262869@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f082b158f4ae97936129732aa9ba0989652153f3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2801 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2869 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2959 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4212 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3240 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2766 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2950 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2909 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2463 "1 flakes 113 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2827 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3237 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2500 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3992 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/732 "2 failures") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2483 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2355 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2488 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2534 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3742 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2882 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2300 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2532 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2486 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/704 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2522 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2703 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1441,7 +1441,8 @@ static Vector<String> authoredGridTrackSizes(Node* node, GridTrackSizingDirectio
     }
 
-    if (!is<CSSValueList>(cssValue))
+    auto* cssValueList = dynamicDowncast<CSSValueList>(*cssValue);
```
Should this be `dynamicDowncast<CSSValueList>(cssValue)` in case `cssValue` is `nullptr` here? I see that we try to handle a scenario where that is the case is above, but is it guaranteed that it can no longer be `nullptr` by the time we get here? 

I think the old code handled that case with `!is<CSSValueList>(cssValue)` so there might be a slight behavior change here?

sammygill replied:
(sorry for the noise! I guess I didn't approve these comments when I first tried to make the review on GitHub)

brentfulgham replied:
It looks like @patrickangle ended up encountering the crash you predicted here, and fixed it in PR [#13260](https://github.com/WebKit/WebKit/pull/13260).

brentfulgham replied:
So it's a shame the comment wasn't approved back in April! :-)

patrickangle replied:
lol all good! I did indeed end up having to fix that crash 😅

patrickangle committed:
Web Inspector: Regression(260175@main) Grid overlay for grids using repeat syntax triggers assert performing downcast
https://bugs.webkit.org/show_bug.cgi?id=255311
rdar://107764654

Reviewed by Devin Rousso.

After 260175@main, CSSGridIntegerRepeatValue and CSSGridAutoRepeatValue no longer inherit from CSSValueList, and instead
inherit from CSSValueContainingVector. The inspector overlay code was written in such a way that we were checking if the
class was one of the repeat types, but then downcasting to CSSValueList making that relation less apparent. Converting
the downcasts to dynamicDowncasts where we previously had the `is` checks fixes the issue and guards against this
mistake in the future.

* LayoutTests/inspector/dom/showGridOverlay.html:
* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::authoredGridTrackSizes):

Canonical link: https://commits.webkit.org/262869@main
